### PR TITLE
replace `FutureService` with Tower's upstream impl

### PR DESCRIPTION
**Pull request by hawkw**
Tower 0.4 included a `FutureService` type which is basically identical
to the version we implemented for converting a `MakeService` into a
`NewService`. This branch replaces our implementation with the upstream
implementation from Tower.

It was necessary to add `MapErr` layers to convert the error types into
boxed errors, since our `FutureService` did that automatically, while
Tower's does not.

Depends on tower-rs/tower#523 from upstream, which is currently
unreleased.

### CODE REVIEW

**Review by olix0r — COMMENTED** [review]
This looks good but I have a slight preference to wait for an actual rather than taking more git deps.

**Comment by olix0r** [comment]
Marking as draft until we have a non-git dep

**Comment by hawkw** [comment]
Tower 0.4.3 has been published, so I've removed the git dep and marked this as ready.

File: linkerd/app/core/src/svc.rs [hunk 1]
```diff
@@ -18,7 +18,11 @@ use tower::{
     make::MakeService,
 };
 pub use tower::{
-    layer::Layer, service_fn as mk, spawn_ready::SpawnReady, util::Either, Service, ServiceExt,
+    layer::Layer,
+    service_fn as mk,
+    spawn_ready::SpawnReady,
+    util::{Either, MapErrLayer},
```
Comment (olix0r) [inline linkerd/app/core/src/svc.rs:24]:
```suggestion
    util::{Either, MapErr},
```

File: linkerd/app/outbound/src/server.rs [hunk 2]
```diff
@@ -146,6 +146,7 @@ where
     let tcp_forward = svc::stack(tcp_connect)
         .push_make_thunk()
         .push_on_response(tcp::Forward::layer())
+        .push(svc::MapErrLayer::new(Into::into))
```
Comment (olix0r) [inline linkerd/app/outbound/src/server.rs:149]:
```suggestion
        .push(svc::MapErr::layer(Into::into))
```

File: linkerd/app/outbound/src/ingress.rs [hunk 3]
```diff
@@ -73,6 +73,7 @@ where
         .into_inner();
 
     svc::stack(http)
+        .push_on_response(svc::MapErrLayer::new(Into::into))
```
Comment (olix0r) [inline linkerd/app/outbound/src/ingress.rs:76]:
```suggestion
        .push_on_response(svc::MapErr::layer(Into::into))
```

File: linkerd/app/outbound/src/http/logical.rs [hunk 4]
```diff
@@ -73,6 +73,7 @@ where
                 .push(svc::FailFast::layer("HTTP Balancer", dispatch_timeout))
                 .push(metrics.stack.layer(stack_labels("http", "concrete"))),
         )
+        .push(svc::MapErrLayer::new(Into::into))
```
Comment (olix0r) [inline linkerd/app/outbound/src/http/logical.rs:76]:
```suggestion
        .push(svc::MapErr::layer(Into::into))
```

**Review by olix0r — APPROVED** [review]

File: Cargo.toml [hunk 5]
```diff
@@ -66,4 +66,4 @@ debug = false
 [patch.crates-io]
 webpki = { git = "https://github.com/linkerd/webpki", branch = "cert-dns-names-0.21", rev = "b2c3bb3" }
 tonic = { git = "https://github.com/hyperium/tonic" }
-tonic-build = { git = "https://github.com/hyperium/tonic" }
+tonic-build = { git = "https://github.com/hyperium/tonic" }
```
Comment (olix0r) [inline Cargo.toml:69]: ?

File: linkerd/app/core/src/svc.rs (hunk 1 above)
Comment (hawkw) [reply linkerd/app/core/src/svc.rs:24]: i actually tried this initially and it didn't compile — `MapErr` is generic over a Service and a fn, while `MapErrLayer` is only generic over the fn type. using the `layer` fn requires a type for the Service type param, which apparently can't be inferred in this usage. :/

File: Cargo.toml (hunk 5 above)
Comment (hawkw) [reply Cargo.toml:69]: wait, what the heck, is this a whitespace change? i can't see what's different...

**Pull request merged**